### PR TITLE
http: new disable flag to disable a filter in the filter chain by default

### PR DESCRIPTION
**Pull request by wbpcode**
Commit Message: http api: new disable flag to disable a filter in the filter chain by default
Additional Description:

In the multiple-tenant gateway, there may be dozen of filters in the http filter chain. And most of routes or users will only use part of these filters. For example, `filter a` for `route 1`, `filter b` for `route 2`.
But in current filter chain implementation, all filters in the chain will be created and iterated, even these filters may do nothing if there are no valid route level configurations.

This problem inspire us to create route level `disable` flag to disable a filter explicitly in specific route/VH. See #29407.

It works pretty well. At least, now we have a way to skip unnecessary filters in the chain and reduce lots of unnecessary CPU overhead.

However, in the production env, we want to disable a filter completely and only enable it when a route level config is provided explicitly.

We can achieve this target by setting disable flag in the `RouteConfiguration.typed_per_filter_config`. But this means at every time a new filter is added, we need update both the LDS and RDS.

I think this thing could be simpler by adding a new flag in the `HttpFilter`. When a new filter is added to the chain, we can specify that it is disabled by default and only enabled it if a valid route configuration is configured explicitly.

Risk Level: low.
Testing: n/a.
Docs Changes: n/a.
Release Notes: n/a.
Platform Specific Features: n/a.

### CODE REVIEW

File: api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto [hunk 1]
```diff
@@ -1167,6 +1167,21 @@ message HttpFilter {
...
+  //
+  // A disabled filter will be treated as enabled in following cases:
+  //
+  // 1. Valid per filter configuration is configured in the ``typed_per_filter_config`` of route config.
+  // 2. :ref:`FilterConfig <envoy_v3_api_msg_config.route.v3.FilterConfig>` is configured in the
+  //    ``typed_per_filter_config`` of route config and the
+  //    :ref:`disabled of FilterConfig <envoy_v3_api_field_config.route.v3.FilterConfig.disabled>`
+  //    is set to false.
+  //
+  // Terminal filters (e.g. ``envoy.filters.http.router``) should never be marked as disabled by default.
```
Comment (mattklein123) [inline api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto:1183]: Can we just make this a config error? So I would just say terminal filters cannot be marked disabled by default.

**Review by mattklein123 — COMMENTED** [review]
Makes sense to me. This should be small enough that I will just wait for the implementation. Thank you.

/wait

File: api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto (hunk 1 above)
Comment (wbpcode) [reply api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto:1183]: Of course. :)

File: api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto [hunk 2]
```diff
@@ -1169,6 +1169,20 @@ message HttpFilter {
...
+  //
+  // A disabled filter will be treated as enabled in following cases:
+  //
+  // 1. Valid per filter configuration is configured in the ``typed_per_filter_config`` of route config.
+  // 2. :ref:`FilterConfig <envoy_v3_api_msg_config.route.v3.FilterConfig>` is configured in the
+  //    ``typed_per_filter_config`` of route config and the
+  //    :ref:`disabled of FilterConfig <envoy_v3_api_field_config.route.v3.FilterConfig.disabled>`
+  //    is set to false.
+  //
+  // Terminal filters (e.g. ``envoy.filters.http.router``) could not be marked as disabled.
```
Comment (mattklein123) [inline api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto:1184]:
```suggestion
  // If true, the filter is disabled by default and must be explicitly enabled by setting
  // per filter configuration in the route configuration.
  //
  // A disabled filter will be treated as enabled in the following cases:
  //
  // 1. Valid per filter configuration is configured in the ``typed_per_filter_config`` of route config.
  // 2. :ref:`FilterConfig <envoy_v3_api_msg_config.route.v3.FilterConfig>` is configured in the
  //    ``typed_per_filter_config`` of route config and the
  //    :ref:`disabled of FilterConfig <envoy_v3_api_field_config.route.v3.FilterConfig.disabled>`
  //    is set to false.
  //
  // Terminal filters (e.g. ``envoy.filters.http.router``) cannot be marked as disabled.
```

File: changelogs/current.yaml [hunk 3]
```diff
@@ -385,6 +385,12 @@ new_features:
   change: |
     Added filters to update the filter state for :ref:`the HTTP requests <config_http_filters_set_filter_state>` and
     :ref:`the TCP connections <config_network_filters_set_filter_state>`.
+- area: http
+  change: |
+    Added :ref:`disabled
+    <envoy_v3_api_field_extensions.filters.network.http_connection_manager.v3.HttpFilter.disabled>`
+    flag to disable a filter in the filter chain by default. This filter could be enabled explicitly by setting
+    valid per filter config in the route configuration.
```
Comment (mattklein123) [inline changelogs/current.yaml:393]:
```suggestion
    Added :ref:`disabled
    <envoy_v3_api_field_extensions.filters.network.http_connection_manager.v3.HttpFilter.disabled>`
    flag to disable a filter in the filter chain by default. The filter can be enabled explicitly by setting
    valid per filter config in the route configuration.
```

File: source/common/http/filter_chain_helper.h [hunk 4]
```diff
@@ -85,12 +90,25 @@ class FilterChainHelper : Logger::Loggable<Logger::Id::config> {
...
+
+    const bool disabled_by_default = proto_config.disabled();
+
+    // Ensure the terminal filter will not be disabled by default. Because the terminal filter must
+    // be the last filter in the chain (ensured by 'validateTerminalFilters') and terminal flag of
+    // dynamic filter could not determined and may changed at runtime, so we check the last filter
+    // flag here as alternative.
+    if (last_filter_in_current_config && disabled_by_default) {
+      return absl::InvalidArgumentError(fmt::format(
+          "Error: the last (terminal) filter ({}) in the chain could not be disabled by default.",
```
Comment (mattklein123) [inline source/common/http/filter_chain_helper.h:102]:
```suggestion
          "Error: the last (terminal) filter ({}) in the chain cannot be disabled by default.",
```

File: test/extensions/filters/http/header_mutation/header_mutation_integration_test.cc [hunk 5]
```diff
@@ -37,6 +37,27 @@ name: donwstream-header-mutation
           key: "downstream-global-flag-header"
           value: "downstream-global-flag-header-value"
         append_action: APPEND_IF_EXISTS_OR_ADD
+)EOF",
+                                 true);
+
+    config_helper_.prependFilter(R"EOF(
+name: donwstream-header-mutation-disabled-by-default
```
Comment (mattklein123) [inline test/extensions/filters/http/header_mutation/header_mutation_integration_test.cc:44]:
```suggestion
name: downstream-header-mutation-disabled-by-default
```

**Review by mattklein123 — CHANGES_REQUESTED** [review]
Thanks LGTM with small nits.

/wait

**Review by mattklein123 — APPROVED** [review]

**Comment by wbpcode** [comment]
@mattklein123 could you give a new stamp? Thanks. I just fixed the CI (fixed the exception message).

**Review by mattklein123 — APPROVED** [review]

**Pull request merged**